### PR TITLE
Fix context tracking with als run

### LINKED ISSUE #3570: Async context is broken for POST requests when its created inside 'onRequest' hook
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.25.1

### Plugin version

5.3.0

### Node.js version

16.10

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.0

### Description

When `next` handler is executed inside `AsyncLocalStorage`, the context gets lost for requests with payload. Post request without a payload will work fine.

Originally discovered in nestjs project, connected issue: https://github.com/nestjs/nest/issues/8837

Note that I had similar issues with express middlewares too and they were caused by the middleware being registered too early, before the body parser. Not sure how that works in fastify, but I would say here the issue will be very similar.

### Steps to Reproduce

storage.ts
```ts
import { AsyncLocalStorage } from 'async_hooks';

export const storage = new AsyncLocalStorage<any>();
export const state = { counter: 0 };
```

main.ts
```
  const fastify = Fastify({ logger: true });
  await fastify.register(middie);
  fastify.use((req: any, res: any, next: any) => {
    const id = state.counter++;
    console.log('In Middleware with id ' + id);
    storage.run({ id }, next);
  });

  fastify.get('/', function (request, reply) {
    const id = storage.getStore()?.id;
    console.log('get', id);
    reply.send({ id });
  });

  fastify.post('/', function (request, reply) {
    const id = storage.getStore()?.id;
    console.log('post', id);
    reply.send({ id });
  });
```

minimal repro here:

https://github.com/B4nan/fastify-middie-als

```
yarn
yarn start
# get request works fine, will print the context ID as a response
curl localhost:3000
# fails to get the ID from ALS if there is actual payload
curl -d '{}' -H 'Content-Type: application/json' localhost:3000
# without the payload it will work fine (even for post request)
curl -X POST localhost:3000
```

### Expected Behavior

The async context should not be broken, requests with payload should work the same as those without it in this manner.

### PR DESCRIPTION
**Pull request by mcollina**
Fixes #3570 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
@B4nan can you please verify this fixes the Nest.js problem?

**Comment by B4nan** [comment]
Amazing! Will verify this first thing in the morning, need to go AFK now.

**Review by zekth — APPROVED** [review]
Implementation is LGTM. Does it have any perf impact? i guess not.

**Comment by mcollina** [comment]
It'd be minimal if at all. Once we drop Node v12 I can remove the 2 closures.

There is not much we can do about it anyway unless it's fixed in Node.js.

**Review by delvedor — APPROVED** [review]
LGTM

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Comment by mittche** [comment]
@mcollina what to do to test?

File: lib/contentTypeParser.js [hunk 1]
```diff
@@ -163,12 +165,15 @@ ContentTypeParser.prototype.run = function (contentType, handler, request, reply
...
 
   function done (error, body) {
-    if (error) {
-      reply.send(error)
-    } else {
-      request.body = body
-      handler(request, reply)
-    }
+    // We cannot use resource.bind() because it is broken in node v12.
+    resource.runInAsyncScope(() => {
```
Comment (kibertoad) [inline lib/contentTypeParser.js:169]: does this introduce noticeable overhead for people who do not use ALS at all?

Comment (mcollina) [reply lib/contentTypeParser.js:169]: I've done a few tests in the past and it was negligible. The major cost is the allocation of an additional object that we throw away after it's all said and done.

**Review by kibertoad — APPROVED** [review]

**Review by climba03003 — APPROVED** [review]
LGTM.

File: lib/contentTypeParser.js [hunk 2]
```diff
@@ -163,12 +165,15 @@ ContentTypeParser.prototype.run = function (contentType, handler, request, reply
   }
 
   function done (error, body) {
-    if (error) {
-      reply.send(error)
-    } else {
-      request.body = body
-      handler(request, reply)
-    }
+    // We cannot use resource.bind() because it is broken in node v12.
```
Comment (jsumners) [inline lib/contentTypeParser.js:168]: Probably should have a link to the issue here, but 🤷‍♂️

**Pull request merged**